dask/dataframe/core.py

```diff
@@ -357,6 +357,24 @@ def _rebuild(self, dsk, *, rename=None):
     def _constructor(self):
         return new_dd_object
 
+    @property
+    def divisions(self):
+        """
+        The "dividing lines" used between the index of the partitions.
+        When ``divisions = (0, 10, 50, 100)``, there are three partitions
+        with indices containing [0, 10), [10, 50), and [50, 100), respectively.
+        When every item is in ``divisions`` is ``None`` the divisions are unknown.
+        Some operations can still be performed, but chaining operations might fail.
+        It is uncommon to set ``divisions`` directly, instead use ``set_index``
+        which sorts and splits the data as needed.
+        See https://docs.dask.org/en/latest/dataframe-design.html#partitions.
+        """
+        return self._divisions
+
+    @divisions.setter
+    def divisions(self, value):
+        self._divisions = tuple(value)
```

Comment:
It _could_ also be worth checking that this is sorted in increasing order? If you set divisions to something un-sorted, you'd probably get some weird errors down the line.

Reply:
That seems like a sensible check - added this in along with a check that unknown divisions are being set properly (since we already need to check that `None` isn't in `divisions` before checking sortedness)

Reply:
Following up here, we might want to make an ascending check a little more nuanced, as it looks like there are some situations where technically non-ascending `divisions` are assigned:

```python
_________________________ test_set_index_categorical __________________________
[gw3] win32 -- Python 3.7.12 C:\Miniconda3\envs\test-environment\python.exe

    def test_set_index_categorical():
        # https://github.com/dask/dask/issues/5671
        order = list(reversed(string.ascii_letters))
        values = list(string.ascii_letters)
        random.shuffle(values)
        dtype = pd.api.types.CategoricalDtype(order, ordered=True)
        df = pd.DataFrame({"A": pd.Categorical(values, dtype=dtype), "B": 1})
    
>       result = dd.from_pandas(df, npartitions=2).set_index("A")

dask\dataframe\tests\test_shuffle.py:908: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dask\dataframe\core.py:4430: in set_index
    **kwargs,
dask\dataframe\shuffle.py:206: in set_index
    df, index, divisions, shuffle=shuffle, drop=drop, compute=compute, **kwargs
dask\dataframe\shuffle.py:325: in set_partition
    df4.divisions = methods.tolist(divisions)
dask\dataframe\core.py:4194: in __setattr__
    object.__setattr__(self, key, value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Dask DataFrame Structure:
                   B
npartitions=2       
               int64
                 ...
                 ...
Dask Name: set_index_post_scalar, 16 tasks
value = ['Z', 'C', 'a']

    @divisions.setter
    def divisions(self, value):
        if None in value:
            if any(v is not None for v in value):
                raise ValueError("divisions must be either all null or all non-null")
        elif any(a > b for a, b in zip(value, value[1:])):
>           raise ValueError("divisions must be sorted in ascending order")
E           ValueError: divisions must be sorted in ascending order

dask\dataframe\core.py:390: ValueError
```

Reply:
I have a feeling it's the latter.

Reply:
After looking into this, think this is actually on us - this failure is coming up when we attempt to set the index to an ordered categorical column. Though the elements of `divisions` are strings, they represent categories with a different ordering.

One way I can imagine this could be resolved is if we checked if `divisions` is sorted when it's cast as a column/index with the same dtype as `self.index`? Or if we can access the ordering from the categorical dtype, we could make an override specifically for that index dtype.